pkg/controller/apiextensions/definer/reconciler.go

```diff
@@ -0,0 +1,275 @@
+	newDefinerFn := func() Definer { return &v1alpha1.InfrastructureDefinition{} }
+	r := &reconciler{
+		client:     kube,
+		Finalizer:  resource.NewAPIFinalizer(kube, finalizerName),
+		crd:        NewAPIInfrastructureClient(kube),
+		mgr:        mgr,
+		newDefiner: newDefinerFn,
+		log:        logging.NewNopLogger(),
+		recorder:   event.NewNopRecorder(),
+	}
+
+	for _, f := range opts {
+		f(r)
+	}
+	// TODO(muvaf): we don't have a use case but it'd be consistent if we allowed
+	// different controller engines to be configured via ReconcilerOption.
+	r.ctrl = composite.NewControllerEngine(mgr, r.log)
+	return r
+}
+
+// ReconcilerOption is used to configure the definer reconciler.
+type ReconcilerOption func(*reconciler)
+
+// WithLogger returns a ReconcilerOption that configures the definer reconciler
+// with given logger.
+func WithLogger(log logging.Logger) ReconcilerOption {
+	return func(r *reconciler) {
+		r.log = log
+	}
+}
+
+// WithRecorder returns a ReconcilerOption that configures the definer reconciler
+// with given event.Recorder.
+func WithRecorder(recorder event.Recorder) ReconcilerOption {
+	return func(r *reconciler) {
+		r.recorder = recorder
+	}
+}
+
+// TODO(muvaf): There are three things that are managed: definer, crd and controller.
+// We should consider having nice sub structs to represent those separately. This
+// will probably be a must when it's used by AppDefinition as well.
+type reconciler struct {
+	client client.Client
+	mgr    manager.Manager
+	ctrl   *composite.ControllerEngine
+	resource.Finalizer
+	crd        Client
+	newDefiner func() Definer
+
+	log      logging.Logger
+	recorder event.Recorder
+}
+
+// Reconcile is the loop function of reconciliation.
+func (r *reconciler) Reconcile(req reconcile.Request) (reconcile.Result, error) { // nolint:gocyclo
+	log := r.log.WithValues("request", req, "gvk", v1alpha1.InfrastructureDefinitionGroupVersionKind.String())
+	log.Debug("Reconciling")
+
+	ctx, cancel := context.WithTimeout(context.Background(), timeout)
+	defer cancel()
+
+	// TODO(muvaf): this will be generic as there will be ApplicationDefinition
+	// type, too.
+	definer := &v1alpha1.InfrastructureDefinition{}
+	if err := r.client.Get(ctx, req.NamespacedName, definer); err != nil {
+		// In case object is not found, most likely the object was deleted and
+		// then disappeared while the event was in the processing queue. We don't
+		// need to take any action in that case.
+		log.Debug(errGetInfraDef, "error", err)
+		return reconcile.Result{}, errors.Wrap(resource.IgnoreNotFound(err), errGetInfraDef)
+	}
+
+	log = log.WithValues(
+		"uid", definer.GetUID(),
+		"version", definer.GetResourceVersion(),
+		"name", definer.GetName(),
+	)
+
+	crd, err := r.crd.Get(ctx, definer)
+	if resource.IgnoreNotFound(err) != nil {
+		log.Debug(errGetCRD, "error", err)
+		return reconcile.Result{RequeueAfter: shortWait}, errors.Wrap(err, errGetCRD)
+	}
+	exists := !kerrors.IsNotFound(err)
```

Comment:
`exists := err != nil` would also work here if we make it this far, right? Not sure whether that would be any more readable, but my brain struggled to parse `!kerrors.IsNotFound` a little due to the double negative (e.g. "is not is not found" rather than "is found").

Reply:
`exists := err != nil` would be non-obvious for readers though even though we skip the `NotFound` above.